rts/Rml/Backends/RmlUi_Backend.cpp

```diff
@@ -0,0 +1,495 @@
+ */
+struct BackendData {
+	RmlSystemInterface system_interface;
+#ifndef HEADLESS
+	RenderInterface_GL3_Recoil render_interface;
+#else
+	RenderInterface_Headless render_interface;
+#endif
+	VFSFileInterface file_interface;
+
+	SDL_Window* window = nullptr;
+	SDL_GLContext gl_context = nullptr;
+	std::vector<Rml::Context*> contexts;
+	InputHandler::SignalType::connection_type inputCon;
+	CRmlInputReceiver inputReceiver;
+
+	// make atomic_bool?
+	bool initialized = false;
+	bool debuggerAttached = false;
+	int winX = 1;
+	int winY = 1;
+
+	lua_State* ls = nullptr;
+	Rml::SolLua::SolLuaPlugin* luaPlugin = nullptr;
+
+	CtxMutex contextMutex;
+	Rml::UniquePtr<PassThroughPlugin> plugin;
+    Rml::UniquePtr<Rml::ElementInstancerGeneric<RmlGui::ElementLuaTexture>> element_lua_texture_instancer;
+};
+
+static Rml::UniquePtr<BackendData> data;
+
+bool RmlInitialized()
+{
+	return data && data->initialized;
+}
+
+bool RmlGui::Initialize(SDL_Window* target_window, SDL_GLContext target_glcontext, int winX,
+                        int winY)
+{
+	data = Rml::MakeUnique<BackendData>();
+
+	if (!data->render_interface) {
+		data.reset();
+		fprintf(stderr, "Could not initialize OpenGL3 render interface.");
+		return false;
+	}
+
+	data->window = target_window;
+	data->gl_context = target_glcontext;
+
+	Rml::SetFileInterface(&data->file_interface);
+	Rml::SetSystemInterface(RmlGui::GetSystemInterface());
+	Rml::SetRenderInterface(RmlGui::GetRenderInterface());
+
+	data->render_interface.SetViewport(winX, winY);
+	data->winX = winX;
+	data->winY = winY;
+
+	Rml::Initialise();
+
+	Rml::LoadFontFace("fonts/FreeSansBold.otf", true);
+	data->inputCon = input.AddHandler(&RmlGui::ProcessEvent);
+	data->initialized = true;
+
+	data->element_lua_texture_instancer = Rml::MakeUnique<Rml::ElementInstancerGeneric<ElementLuaTexture>>();
+	Rml::Factory::RegisterElementInstancer("texture", data->element_lua_texture_instancer.get());
+
+	data->plugin = Rml::MakeUnique<PassThroughPlugin>(OnContextCreate, OnContextDestroy);
+	Rml::RegisterPlugin(data->plugin.get());
+
+	return true;
+}
+
+bool RmlGui::InitializeLua(lua_State* lua_state)
+{
+	if (!RmlInitialized()) {
+		return false;
+	}
+	sol::state_view lua(lua_state);
+	data->ls = lua_state;
+	data->luaPlugin = Rml::SolLua::Initialise(&lua, "rmlDocumentId");
+	data->system_interface.SetTranslationTable(&data->luaPlugin->translationTable);
+	return true;
+}
+
+bool RmlGui::RemoveLua()
+{
+	if (!RmlInitialized() || !data->ls) {
+		return false;
+	}
+	data->luaPlugin->RemoveLuaItems();
+	Update();
+	Rml::UnregisterPlugin(data->luaPlugin);
+	data->system_interface.SetTranslationTable(nullptr);
+	data->luaPlugin = nullptr;
+	data->ls = nullptr;
+
+	return true;
+}
+
+void RmlGui::Shutdown()
+{
+	if (!RmlInitialized()) {
+		return;
+	}
+
+	Rml::UnregisterPlugin(data->plugin.get());
+	data.reset();
+
+	// removes all contexts
+	Rml::Shutdown();
+}
+
+void RmlGui::Reload()
+{
+	if (!RmlInitialized()) {
+		return;
+	}
```

Comment:
Nitpick: For this specific function, is this the desired behaviour? Like, if RmlUi has not been initialized yet, should calling `Reload()` cause initialization instead of nothing? I am asking because `RmlGui::Reload()` might be used here:

https://github.com/beyond-all-reason/spring/pull/1415/files#diff-5976529b058f061411633421ef38f6f3952fad64cc5c5ce50b0b66a31c1b6f45R154-R172

https://github.com/beyond-all-reason/spring/blob/87fc77eab84e0f1b227a46ee91c4282747c75719/rts/Game/UI/GuiHandler.cpp#L154-L172

And I do not know if that can ever be called in typical scenarios without having RmlUi initialized. That functions name is `EnableLuaUI()`, which sounds a bit like it might be called without things being initialized first (if so, can those things include RmlUi?).

Reply:
This has been resolved :+1: .